pkg/runtime/pubsub/bulksub.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pubsub
+
+import (
+	"context"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/dapr/dapr/utils"
+
+	contribPubsub "github.com/dapr/components-contrib/pubsub"
+)
+
+const (
+	maxBulkCountKey     string = "maxBulkCount"
+	maxBulkLatencyMsKey string = "maxBulkLatencyMilliSeconds"
+
+	defaultMaxBulkCount     int = 100
+	defaultMaxBulkLatencyMs int = 1 * 1000
+)
+
+// msgWithCallback is a wrapper around a message that includes a callback function
+// that is called when the message is processed.
+type msgWithCallback struct {
+	msg contribPubsub.BulkMessageEntry
+	cb  func(error)
+}
+
+// defaultBulkSubscriber is the default implementation of BulkSubscriber.
+// It is used when the component does not implement BulkSubscriber.
+type defaultBulkSubscriber struct {
+	p contribPubsub.PubSub
+}
+
+// NewDefaultBulkSubscriber returns a new defaultBulkSubscriber from a PubSub.
+func NewDefaultBulkSubscriber(p contribPubsub.PubSub) *defaultBulkSubscriber {
+	return &defaultBulkSubscriber{
+		p: p,
+	}
+}
+
+// BulkSubscribe subscribes to a topic using a BulkHandler.
+// Dapr buffers messages in memory and calls the handler with a list of messages
+// when the buffer is full or max await duration is reached.
+func (p *defaultBulkSubscriber) BulkSubscribe(ctx context.Context, req contribPubsub.SubscribeRequest, handler contribPubsub.BulkHandler) error {
+	cfg := contribPubsub.BulkSubscribeConfig{
+		MaxBulkCount:               utils.GetIntOrDefault(req.Metadata, maxBulkCountKey, defaultMaxBulkCount),
+		MaxBulkLatencyMilliSeconds: utils.GetIntOrDefault(req.Metadata, maxBulkLatencyMsKey, defaultMaxBulkLatencyMs),
+	}
+
+	msgCbChan := make(chan msgWithCallback, cfg.MaxBulkCount)
+	go processBulkMessages(ctx, req.Topic, msgCbChan, cfg, handler)
+
+	// Subscribe to the topic and listen for messages.
+	return p.p.Subscribe(ctx, req, func(ctx context.Context, msg *contribPubsub.NewMessage) error {
+		var err error
+		done := make(chan struct{})
+
+		bulkMsg := contribPubsub.BulkMessageEntry{
+			EntryID:  uuid.NewString(),
```

Comment:
It may be better to use NewRandom() instead, as NewString() or New().String() may panic, if not enough entropy. One quick ref that I could find: https://github.com/containerish/OpenRegistry/issues/125

Reply:
Great catch, changing it to `NewRandom`.